feat: user search by email feature

Benderbej opened the pull request:
EPMLSTRCMWP-110: Search Users by email feature

Please, pay attention: this request contains - TODO notes at cromwell.pipeline.datastorage.dao.entry.UserEntry 
this must be reviewed

myazinn commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -0,0 +1,32 @@
+    }
+
+  private def getSeqJson(seq: Seq[UserResponse]): String = {
+    Json.stringify(Json.toJson(seq))
```
Why do you need to return a string instead of `json`?
Especially since this can be done via implicit conversions without the need for explicit method calls.

Benderbej replied:
I've just trying to compose parts of my code together, to work with them. That was my minimum task to do. Unfortunatlly I have not enough experience to do it right way, right now. But. I am going to do this now

Benderbej replied:
The problem is that Seq[UserResponse] highlighted as UNmarshallable, but I have all the implicits for UserResponsible (I've read that I do not anything else to make sequence marshallable, & all I have to do is to make my object marshallable, - sequance will be marshallable automatically by play's lib) and I do not know what to do

Benderbej replied:
complete method in controller requires ToResponseMarshallable as a return type, but JsValue is not this type. 2) When I use implicits (both Formats & writes()/reads()) compiler do not recognise UserResponse object as mershallable, but if i use String as a return type & writes() for example, I see that play use this mathod(writes()) to write my String...

myazinn commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -0,0 +1,32 @@
+          val res: Future[Seq[UserResponse]] = userService.getUsersByEmail(email)
+          onComplete(res) {
+            case Success(r)                            => complete(getSeqJson(r))
+            case Failure(exception)                    => complete(StatusCodes.InternalServerError)
```
No need to give a name for a variable if you are not using it. You can use `_`

Benderbej replied:
have done

myazinn commented on portal/src/main/scala/cromwell/pipeline/datastorage/dao/entry/UserEntry.scala:
```diff
@@ -29,6 +29,12 @@ class UserEntry(val pipelineDatabaseEngine: PipelineDatabaseEngine) {
-  def addUserAction(user: User) = (users.returning(users.map(_.userId))) += user
+  def getUsersByEmailAction(emailPattern: String) = {
+    val query = for {//TODO - can we make it compiled?
+      user <- users if user.email like "%"+emailPattern+"%" //TODO - is it safe to give parameters to DB this way(to prevent SQL-injections)? what syntax should we use to concatenate strings? any preferred formaters?
```
I'm not sure about Slick, but in general, this is not safe.
To format string you can use [string interpolation](https://docs.scala-lang.org/overviews/core/string-interpolation.html) "%$emailPattern%"

Benderbej replied:
users.filter(_.email like s"%$emailPattern%").result 
This works fine, thank you so much

myazinn commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -0,0 +1,32 @@
+  private def getSeqJson(seq: Seq[UserResponse]): String = {
+    Json.stringify(Json.toJson(seq))
+  }
+}
```
Turn on ensure line feed option in the Intellij Idea to get rid of this git warning.

Benderbej replied:
done

Benderbej commented:
Its littlebit strange, but I made some changes in CromwellPipelineApp file, then squashed my commits into one, then forced-pushed it to repo, but git answered me that "everything up-to-date", while my local CromwellPipelineApp still differ from that is origin

myazinn commented on portal/src/main/scala/cromwell/pipeline/datastorage/dto/user/UserResponse.scala:
```diff
@@ -0,0 +1,18 @@
+import cromwell.pipeline.datastorage.dto.{ User, UserId }
+import play.api.libs.json.{ Format, JsPath, Json, OFormat, Reads, Writes }
+
+final case class UserResponse(user: User)
```
I have no idea. Totally agree with you, in that case there is no need for additional wrapper.

Benderbej commented on portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala:
```diff
@@ -26,7 +26,8 @@ object CromwellPipelineApp extends App {
 
-  val route = authController.route ~ securityDirective.authenticated { _ =>
+  val route = authController.route ~ userController.route ~ securityDirective.authenticated { _ =>
+    //userController.route
```
quite sure, we do not, I have tryied to remove it but had some problems with force-pushing, and then need to correct and add more thing, so I desided to postpone this correction

Benderbej commented on portal/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala:
```diff
@@ -5,6 +5,7 @@ import akka.http.scaladsl.model.{ HttpEntity, StatusCodes }
 import akka.http.scaladsl.testkit.ScalatestRouteTest
 import cromwell.pipeline.controller.AuthController._
 import cromwell.pipeline.datastorage.dto.auth.{ AuthResponse, SignInRequest, SignUpRequest }
+//import cromwell.pipeline.datastorage.dto.auth.{ AuthResponse, SignUpRequest }
```
have done

Benderbej commented on portal/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala:
```diff
@@ -0,0 +1,43 @@
+
+      "return the sequence of users" taggedAs (Controller) in {
+        val userByEmailRequest: String = "@mail"
+        val userByEmailResponse: Seq[UserResponse] = uEmailRespSeq
```
I didnt know why did I do this. May be I was exhausted and lost my attention, in any case, Corrected!

Benderbej commented on portal/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala:
```diff
@@ -0,0 +1,43 @@
+
+        Get("/users?email=" + userByEmailRequest) ~> userController.route ~> check {
+          status shouldBe StatusCodes.OK
+          responseAs[JsValue] shouldEqual Json.toJson(userByEmailResponse)
```
"responseAs[Seq[UserResponse]] shouldEqual uEmailRespSeq" - this works fine for me, but why didnt we use JsValue and toJson()? This is Controller-test, why dont we deal with json-transformations?

Benderbej replied:
ok, thank you for explaination! removed redundant code, corrected with responseAs[Seq[UserResponse]] shouldEqual uEmailRespSeq

Benderbej commented on portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala:
```diff
@@ -0,0 +1,44 @@
+
+  private val userRepository: UserRepository = stub[UserRepository]
+  private val authUtils: AuthUtils = stub[AuthUtils]
+  private val userId = new UserId("")
```
have done

Benderbej commented on portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala:
```diff
@@ -0,0 +1,44 @@
+  private val userRepository: UserRepository = stub[UserRepository]
+  private val authUtils: AuthUtils = stub[AuthUtils]
+  private val userId = new UserId("")
+  private val picture = new ProfilePicture(Array())
```
have done

Benderbej commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -0,0 +1,31 @@
+    path("users") {
+      get {
+        parameter('email.as[String]) { email =>
+          val res: Future[Seq[UserResponse]] = userService.getUsersByEmail(email)
```
corrected

myazinn commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -0,0 +1,32 @@
+            case Failure(_) =>
+              complete(
+                StatusCodes.InternalServerError,
+                "may be, database error or query syntax error: try to check your request email-pattern"
```
In general, is it a great idea to provide meaningful description. However, as I said earlier during our meeting, the `Controller` layer should not even know that the database exists.

Benderbej replied:
exc.getMessage added, thank you for advice

myazinn commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -0,0 +1,32 @@
+              complete(
+                StatusCodes.InternalServerError,
+                "may be, database error or query syntax error: try to check your request email-pattern"
+              ) //TODO - add text description for this error
```
Remove all comments in the PR if we don't need to see them.

Benderbej replied:
removed

myazinn commented on portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala:
```diff
@@ -12,5 +17,27 @@ final case class User(
 final case class UserId(value: String) extends MappedTo[String]
+
+object UserId {
+  implicit lazy val userIdFormat: OFormat[UserId] = Json.format[UserId]
```
`UserId` should be formatted differently. You can find how this should be done at [this](https://github.com/EpamLifeSciencesTeam/cmwl_pipeline/pull/16) PR.

Benderbej replied:
replaced with  ` implicit lazy val userIdFormat: Format[UserId] = implicitly[Format[String]].inmap(UserId.apply, _.value)` just as how in that PR

myazinn commented on portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala:
```diff
@@ -12,5 +17,27 @@ final case class User(
 final case class ProfilePicture(value: Array[Byte]) extends MappedTo[Array[Byte]]
+
+object ProfilePicture {
+  implicit lazy val profilePictureFormat: OFormat[ProfilePicture] = Json.format[ProfilePicture]
```
Since you have custom `Reads` and `Writes`, you do not need this `Format`

Benderbej replied:
removed, thanks for explaination

Benderbej commented:
Bug was finded. Integration tests do not work in parallel: This string: `parallelExecution in IntegrationTest := false`, in build.sbt make tests succed, while starting it:test

ilyaonishenko commented on portal/src/test/scala/cromwell/pipeline/spec/ServiceSpec.scala:
```diff
@@ -0,0 +1,7 @@
+import org.scalatest.{ AsyncWordSpec, Matchers }
+import org.scalatestplus.mockito.MockitoSugar
+
+abstract class ServiceSpec extends AsyncWordSpec with Matchers with MockitoSugar with ScalaFutures
```
And without `ServiceSpec` as well.

Benderbej replied:
done

Benderbej commented on portal/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala:
```diff
@@ -1,25 +1,69 @@
+        val accessToken = AccessTokenContent(userId.value)
+        when(userService.getUsersByEmail(usersByEmailRequest)).thenReturn(Future.successful(uEmailRespSeq))
+
+        Get("/users?email=" + usersByEmailRequest) ~> userController.route(accessToken) ~> check {
```
I am not quite sure, but I have got it from here: https://doc.akka.io/docs/akka-http/current/routing-dsl/testkit.html

ilyaonishenko replied:
Looks like it is typical way to test queries in Akka-http.
Let's leave it as it is then

DenisVerkhoturov commented:
Please, delete [`sbt` file](https://github.com/Benderbej/cmwl_pipeline/blob/910a9143f922926523068cc21909a3d7682b8075/sbt) you are adding in the root of the project.

Benderbej commented on portal/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala:
```diff
@@ -1,25 +1,69 @@
 import scala.concurrent.Future
 
-class UserControllerTest extends AsyncWordSpec with Matchers with MockitoSugar with ScalatestRouteTest {
+class UserControllerTest extends ControllerSpec {
```
done

Benderbej commented on portal/src/test/scala/cromwell/pipeline/datastorage/dao/repository/UserRepositoryTest.scala:
```diff
@@ -50,6 +51,15 @@ class UserRepositoryTest extends AsyncWordSpec with Matchers with BeforeAndAfter
+
+        val newUser: User = TestUserUtils.getDummyUser()
+        userRepository.addUser(newUser)
+        userRepository.getUsersByEmail(newUser.email).map(repoResp => repoResp shouldBe Vector(newUser))
```
have done this

Benderbej commented on portal/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala:
```diff
@@ -1,25 +1,69 @@
+      "return the sequence of users" taggedAs (Controller) in {
+        val usersByEmailRequest: String = "@mail"
+        val dummyUser: User = TestUserUtils.getDummyUser()
+        val userId = dummyUser.userId
```
removed this redundant val

Benderbej commented on portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala:
```diff
@@ -14,9 +23,14 @@ final case class User(
+  require(DomainValidation.checkEmail(email), EmailDoesNotMeetCriteria.errorMessage)
+  require(DomainValidation.checkFirstName(firstName), FirstNameHasSpecialCharacters.errorMessage)
+  require(DomainValidation.checkLastName(lastName), LastNameHasSpecialCharacters.errorMessage)
+}
```
deleted all code related to this requires

Benderbej commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -1,26 +1,42 @@
+                  StatusCodes.InternalServerError,
+                  exc.getMessage
+                )
+            }
```
have done this

Benderbej commented:
> Please, delete [`sbt` file](https://github.com/Benderbej/cmwl_pipeline/blob/910a9143f922926523068cc21909a3d7682b8075/sbt) you are adding in the root of the project.

deleted and removed from git index

myazinn commented on portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala:
```diff
@@ -14,9 +17,10 @@ final case class User(
   profilePicture: Option[ProfilePicture] = None,
   active: Boolean = true
-)
+) {}
```
Remove this braces

Benderbej replied:
done

myazinn commented on portal/src/main/scala/cromwell/pipeline/service/UserService.scala:
```diff
@@ -1,12 +1,20 @@
+//  def getUsersByEmail(emailPattern: String): Future[Seq[User]] = emailPattern match {
+//    case Failure(exc) => Failure(exc)
+//    case _            => userRepository.getUsersByEmail(emailPattern)
+//  }
```
Remove useless commented lines

Benderbej replied:
done

myazinn commented on portal/src/test/scala/cromwell/pipeline/datastorage/dao/repository/UserRepositoryTest.scala:
```diff
@@ -50,6 +51,16 @@ class UserRepositoryTest extends AsyncWordSpec with Matchers with BeforeAndAfter
+        userRepository
+          .addUser(newUser)
+          .flatMap(
+            _ => userRepository.getUsersByEmail(newUser.email).map(repoResp => repoResp shouldBe Vector(newUser))
```
Checking that the method returns a `Vector` is not a good idea since this is not guaranteed and may cause tests to fail if the method will return, for example, a `List`. It is better to use things like `should contain theSameElementsAs`.

Benderbej replied:
As I got it, I can use - `repoResp should contain theSameElementsAs Vector(newUser)` - instead, Isn't it?

myazinn commented on portal/src/test/scala/cromwell/pipeline/project/build.properties:
```diff
@@ -0,0 +1 @@
+sbt.version=1.3.6
```
What is this?

Benderbej replied:
I wonder too, maybe misclick, removed

myazinn commented on portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala:
```diff
@@ -27,3 +31,16 @@ object UserId {
+  }
+
+  lazy val profilePictureReads: Reads[ProfilePicture] =
+    (JsPath \ "value").read(Reads.of[Array[Byte]]).map(ProfilePicture.apply _)
```
I've just noticed that this is the wrong place to hide `ProfilePicture` content. The content of the profile picture is an array, it is strange to replace it with `None`. You must provide custom `Reads` and `Writes` for a `User` class in order to always write `profilePicture` as `None`.
Also, why do you need to do it right now? :)

Benderbej replied:
done, thanks for explanation

myazinn commented on portal/src/test/scala/cromwell/pipeline/utils/auth/TestUserUtils.scala:
```diff
@@ -11,15 +11,54 @@ object TestUserUtils {
+    )
+  }
+
+  def createDummyUser(
```
Weird.
This method just creates a user, but for some reason does not use the active parameter. Why would anyone need such a method?

Benderbej replied:
done, that's right. it was waste methods.

myazinn commented on portal/src/test/scala/cromwell/pipeline/utils/auth/TestUserUtils.scala:
```diff
@@ -11,15 +11,54 @@ object TestUserUtils {
+      "LastName",
+      active
+    )
+  }
```
This is basically a ctrl+c ctrl+v from `getDummyUser` method. Can you refactor it somehow so one of this method uses another?

Benderbej replied:
removed, done

ilyaonishenko commented on portal/src/main/scala/cromwell/pipeline/controller/UserController.scala:
```diff
@@ -1,26 +1,39 @@
 import scala.util.{ Failure, Success }
 
-class UserController(userService: UserService)(implicit executionContext: ExecutionContext) {
+class UserController(userService: UserService)(implicit executionContext: ExecutionContext) extends PlayJsonSupport {
```
I guess in this case it is better to have composition instead of inheritance.

Benderbej replied:
solved it with `import de.heikoseeberger.akkahttpplayjson.PlayJsonSupport._`

ilyaonishenko commented on portal/src/test/scala/cromwell/pipeline/datastorage/dao/repository/UserRepositoryTest.scala:
```diff
@@ -50,6 +51,19 @@ class UserRepositoryTest extends AsyncWordSpec with Matchers with BeforeAndAfter
+            _ =>
+              userRepository
+                .getUsersByEmail(newUser.email)
+                .map(repoResp => repoResp should contain theSameElementsAs Vector(newUser))
```
1. Please introduce value at line 61.
2. Why do you use Vector here?

Benderbej replied:
1. Why do I need it here, if I do not use it anywhere? 
2. changed type to Seq

myazinn commented on portal/src/test/scala/cromwell/pipeline/utils/auth/TestUserUtils.scala:
```diff
@@ -2,24 +2,31 @@ package cromwell.pipeline.utils.auth
+      "FirstName",
+      "LastName",
+      None,
+      active
```
Include other `lastName` as a parameter too and don't forget to use it in a `User` constructor. Also, rename `emailOpt` to `email`.

Benderbej replied:
done: 
a) added parameters to constructor, 
b) used this params in constructor in getDummyUser(), 
c)renamed emalopt
